dali/operators/generic/histogram/histogram.cc

```diff
@@ -0,0 +1,685 @@
+
+  // Prepare output tensors shapes
+  TensorListShape<> output_shapes, bin_shapes = GetBinShapes(input_shapes.num_samples());
+
+  if (!non_reduction_axes_.empty()) {
+    auto non_reduction_axes_shape = input_shapes.first(non_reduction_axes_.size());
+    output_shapes.resize(input_shapes.num_samples(), non_reduction_axes_.size() + hist_dim_);
+
+    for (int i = 0; i < input_shapes.num_samples(); ++i) {
+      auto out_sh = shape_cat(non_reduction_axes_shape[i], bin_shapes[i]);
+      output_shapes.set_tensor_shape(i, out_sh);
+    }
+  } else {
+    output_shapes = std::move(bin_shapes);
+  }
+
+  // Simplify shapes so histogram of per reduction can be calculated easily.
+  SubdivideTensorsShapes(input_shapes, output_shapes, output_desc);
+}
+
+void HistogramCPU::SubdivideTensorsShapes(const TensorListShape<> &input_shapes,
+                                          const TensorListShape<> &output_shapes,
+                                          OutputDesc &output_desc) {
+  SmallVector<std::pair<int, int>, 2> in_collapse_groups, out_collapse_groups;
+  TensorListShape<> norm_inputs, norm_outputs;
+
+  int reduced_start = 1;
+  if (non_reduction_axes_.empty()) {
+    // Add unitary outer dimension for unfold
+    norm_inputs.resize(input_shapes.num_samples(), input_shapes.sample_dim() + 1);
+    norm_outputs.resize(output_shapes.num_samples(), output_shapes.sample_dim() + 1);
+    for (int i = 0; i < input_shapes.num_samples(); ++i) {
+      norm_inputs.set_tensor_shape(i, shape_cat(1, input_shapes[i]));
+      norm_outputs.set_tensor_shape(i, shape_cat(1, output_shapes[i]));
+    }
+  } else {
+    if (non_reduction_axes_.size() != 1) {
+      // Add collapse group to collapse outer (non-reduction) dimensions
+      in_collapse_groups.push_back(std::make_pair(0, non_reduction_axes_.size()));
+      out_collapse_groups.push_back(std::make_pair(0, non_reduction_axes_.size()));
+      reduced_start = non_reduction_axes_.size();
+    }
+    norm_inputs = input_shapes;
+    norm_outputs = output_shapes;
+  }
+
+  if (reduction_axes_.size() > 1) {
+    // Add collapse group to collapse inner (reduction) dimensions, possibly ommiting channels
+    // dimension.
+    in_collapse_groups.push_back(std::make_pair(reduced_start, reduction_axes_.size()));
+  }
+
+  norm_inputs = collapse_dims(norm_inputs, in_collapse_groups);
+  norm_outputs = collapse_dims(norm_outputs, out_collapse_groups);
+
+  auto splited_input_shapes = unfold_outer_dim(norm_inputs);
+  auto splited_output_shapes = unfold_outer_dim(norm_outputs);
+
+  std::vector<int> split_mapping;
+  split_mapping.reserve(splited_input_shapes.num_samples());
+  auto norm_non_reduced = norm_inputs.first(1);
+  for (int i = 0; i < norm_non_reduced.num_samples(); ++i) {
+    for (int j = 0; j < norm_non_reduced[i][0]; ++j) {
+      split_mapping.push_back(i);
+    }
+  }
+
+  split_mapping_ = std::move(split_mapping);
+  splited_input_shapes_ = std::move(splited_input_shapes);
```

Comment:
"Split" is an irregular verb. split/split/split.
```suggestion
  split_input_shapes_ = std::move(split_input_shapes);
```